In [10]:
# File name: Exercise 4.7 - Data Wrangling
# Author: Sam Abrams
# Created: 11/3/24
# Description: This notebook contains exercises related to the creation of new variables and columns, namely the busiest days of the week, businest hours of the day, and price ranges.

# Practice Work for Exercise 4.7 - Deriving New Variables

In [2]:
# import libraries
import pandas as pd
import numpy as np
import os

In [4]:
# import dataframes
df_ord_prod_merge = pd.read_pickle('/Users/samabrams/Data Analysis Projects/Instacart Basket Analysis/02 Data/Prepared Data/orders_products_merged.pkl')

In [6]:
df_ord_prod_merge

product_id                product_name  aisle_id  department_id  \
0                  1  Chocolate Sandwich Cookies        61             19   
1                  1  Chocolate Sandwich Cookies        61             19   
2                  1  Chocolate Sandwich Cookies        61             19   
3                  1  Chocolate Sandwich Cookies        61             19   
4                  1  Chocolate Sandwich Cookies        61             19   
...              ...                         ...       ...            ...   
32404854       49688      Fresh Foaming Cleanser        73             11   
32404855       49688      Fresh Foaming Cleanser        73             11   
32404856       49688      Fresh Foaming Cleanser        73             11   
32404857       49688      Fresh Foaming Cleanser        73             11   
32404858       49688      Fresh Foaming Cleanser        73             11   

          prices  order_id  user_id  order_number  order_day_of_week  \
0            5.8   3139998      138            28                  6   
1            5.8   1977647      138            30                  6   
2            5.8    389851      709             2                  0   
3            5.8    652770      764             1                  3   
4            5.8   1813452      764             3                  4   
...          ...       ...      ...           ...                ...   
32404854    13.5   1788356   200215             2                  0   
32404855    13.5   3401313   200377             1                  4   
32404856    13.5    809510   200873             5                  3   
32404857    13.5   2359893   200873             9                  3   
32404858    13.5   2385091   205926            11                  1   

          order_hour_of_day  days_since_last_order  add_to_cart_order  \
0                        11                    3.0                  5   
1                        17                   20.0                  1   
2                        21                    6.0                 20   
3                        13                    NaN                 10   
4                        17                    9.0                 11   
...                     ...                    ...                ...   
32404854                  9                    5.0                 27   
32404855                 11                    NaN                  5   
32404856                  8                   15.0                 12   
32404857                 15                    5.0                 11   
32404858                 15                    6.0                 35   

          reordered _merge source  
0                 0   both   both  
1                 1   both   both  
2                 0   both   both  
3                 0   both   both  
4                 1   both   both  
...             ...    ...    ...  
32404854          0   both   both  
32404855          0   both   both  
32404856          0   both   both  
32404857          1   both   both  
32404858          0   both   both  

[32404859 rows x 15 columns]

In [44]:
df = df_ord_prod_merge[:1000000]

In [42]:
# Limits to 1 million rows
df = df_ord_prod_merge[:1000000]

In [32]:
# Defines tags for price ranges
def price_label(row):
    if row['prices'] <= 5:
        return 'low-range product'
    elif (row['prices'] > 5) and (row['prices'] <= 15):
        return 'mid-range product'
    elif (row['prices'] >15):
        return 'high-range product'
    else: return np.nan

In [ ]:
# apply the price_range function
df['price_range'].value_counts(dropna = False)

In [50]:
df['prices'].max()

24.5

In [52]:
# high-range product dataframe
df_high_range = df[df['price_range'] == 'high-range product']

In [54]:
df_high_range.shape

(9344, 16)

In [56]:
df_high_range.head()

product_id       product_name  aisle_id  department_id  prices  \
5893          17  Rendered Duck Fat        35             12    17.1   
5894          17  Rendered Duck Fat        35             12    17.1   
5895          17  Rendered Duck Fat        35             12    17.1   
5896          17  Rendered Duck Fat        35             12    17.1   
5897          17  Rendered Duck Fat        35             12    17.1   

      order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
5893   2045631     6577            25                  3                  9   
5894    425469    11710             6                  5                 15   
5895   3361237    13130            29                  4                 13   
5896    585724    25271             2                  0                 19   
5897   1161642    36912            35                  6                 19   

      days_since_last_order  add_to_cart_order  reordered _merge source  \
5893                    9.0                 12          0   both   both   
5894                   22.0                  2          0   both   both   
5895                    2.0                  4          0   both   both   
5896                   30.0                  7          0   both   both   
5897                   12.0                  2          0   both   both   

             price_range  
5893  high-range product  
5894  high-range product  
5895  high-range product  
5896  high-range product  
5897  high-range product

In [60]:
df_high_range['prices'].max()

24.5

In [66]:
df_high_range.describe()

product_id     aisle_id  department_id       prices      order_id  \
count  9344.000000  9344.000000    9344.000000  9344.000000  9.344000e+03   
mean   1003.401969   100.340325      12.024401    20.886612  1.707355e+06   
std     421.798563    22.417639       0.154298     1.637864  9.864509e+05   
min      17.000000     7.000000      12.000000    15.100000  3.580000e+02   
25%     907.000000   106.000000      12.000000    20.000000  8.457418e+05   
50%    1117.000000   106.000000      12.000000    21.600000  1.717306e+06   
75%    1117.000000   106.000000      12.000000    21.600000  2.562837e+06   
max    2005.000000   122.000000      13.000000    24.500000  3.420357e+06   

             user_id  order_number  order_day_of_week  order_hour_of_day  \
count    9344.000000   9344.000000        9344.000000        9344.000000   
mean   103481.182149     13.551798           2.842787          13.351777   
std     59768.730145     15.056229           2.118507           4.265645   
min        61.000000      1.000000           0.000000           0.000000   
25%     50535.000000      4.000000           1.000000          10.000000   
50%    103328.000000      8.000000           3.000000          13.000000   
75%    155480.750000     18.000000           5.000000          16.000000   
max    206201.000000     99.000000           6.000000          23.000000   

       days_since_last_order  add_to_cart_order    reordered  
count            8550.000000        9344.000000  9344.000000  
mean               13.263275           7.423801     0.521725  
std                 9.655720           6.697112     0.499555  
min                 0.000000           1.000000     0.000000  
25%                 6.000000           3.000000     0.000000  
50%                10.000000           5.000000     1.000000  
75%                21.000000          10.000000     1.000000  
max                30.000000          64.000000     1.000000

## If-statements with loc() Function

In [75]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [71]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [73]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [77]:
df['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     652638
Low-range product     338018
High-range product      9344
Name: count, dtype: int64

Let’s walk through each step of what this is doing. You already know that the loc() function locates a particular column in the dataframe it’s been assigned to. Now, a logical operator (smaller than, larger than, equal to, etc.) is being added to the function to create a condition. The difference here is that there’s no explicit if in your if-statement. Instead, it’s all been implied.

Take a closer look at the first line of code:

df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'
From left to right, the loc() function is being called on the df dataframe. And within the brackets, the values in the “prices” column of the df dataframe are being compared to a value, 15, using the > operator. In normal language, you could say, “if the values in the ‘prices’ column of the df dataframe are greater than 15.”

After the comma comes the implied “then.” Here, a new column called “price_range_loc” is being set equal to the string “High-range product.” This is the same as the label you created in your user-defined function.

The two implied halves of the if-statement, then, would be:

if = df.loc[df['prices'] > 15,
then = 'price_range_loc'] = 'High-range product'
Remember—the comma is key! It’s what separates the “if” from the “then.”

In [80]:
# Using loc() function on orders_products_merged dataframe

In [82]:
df_ord_prod_merge.loc[df_ord_prod_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [84]:
df_ord_prod_merge.loc[(df_ord_prod_merge['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [85]:
df_ord_prod_merge.loc[df_ord_prod_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [88]:
df_ord_prod_merge['price_range_loc'].value_counts(dropna = False)

price_range_loc
NaN                   21208222
Low-range product     10126321
Mid-range product       652638
High-range product      417678
Name: count, dtype: int64

### I think I messed up, that's a lot of NaNs...

## If-Statements with For-Loops

In [93]:
df_ord_prod_merge['order_day_of_week'].value_counts(dropna = False)

order_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [147]:
result = []

for value in df_ord_prod_merge['order_day_of_week']:
    if value == 0:
        result.append('busiest day')
    elif value == 4:
        result.append('least busy day')
    else:
        result.append('regular day')

In [97]:
print('result')

result


In [99]:
print(result)

IOPub data rate exceeded.
The Jupyter server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--ServerApp.iopub_data_rate_limit`.

Current values:
ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
ServerApp.rate_limit_window=3.0 (secs)



In [149]:
df_ord_prod_merge['busiest day'] = result

In [151]:
df_ord_prod_merge['busiest day'].value_counts(dropna = False)

busiest day
regular day       22416875
busiest day        6204182
least busy day     3783802
Name: count, dtype: int64

In [105]:
df_ord_prod_merge

product_id                product_name  aisle_id  department_id  \
0                  1  Chocolate Sandwich Cookies        61             19   
1                  1  Chocolate Sandwich Cookies        61             19   
2                  1  Chocolate Sandwich Cookies        61             19   
3                  1  Chocolate Sandwich Cookies        61             19   
4                  1  Chocolate Sandwich Cookies        61             19   
...              ...                         ...       ...            ...   
32404854       49688      Fresh Foaming Cleanser        73             11   
32404855       49688      Fresh Foaming Cleanser        73             11   
32404856       49688      Fresh Foaming Cleanser        73             11   
32404857       49688      Fresh Foaming Cleanser        73             11   
32404858       49688      Fresh Foaming Cleanser        73             11   

          prices  order_id  user_id  order_number  order_day_of_week  \
0            5.8   3139998      138            28                  6   
1            5.8   1977647      138            30                  6   
2            5.8    389851      709             2                  0   
3            5.8    652770      764             1                  3   
4            5.8   1813452      764             3                  4   
...          ...       ...      ...           ...                ...   
32404854    13.5   1788356   200215             2                  0   
32404855    13.5   3401313   200377             1                  4   
32404856    13.5    809510   200873             5                  3   
32404857    13.5   2359893   200873             9                  3   
32404858    13.5   2385091   205926            11                  1   

          order_hour_of_day  days_since_last_order  add_to_cart_order  \
0                        11                    3.0                  5   
1                        17                   20.0                  1   
2                        21                    6.0                 20   
3                        13                    NaN                 10   
4                        17                    9.0                 11   
...                     ...                    ...                ...   
32404854                  9                    5.0                 27   
32404855                 11                    NaN                  5   
32404856                  8                   15.0                 12   
32404857                 15                    5.0                 11   
32404858                 15                    6.0                 35   

          reordered _merge source        price_range    price_range_loc  \
0                 0   both   both  mid-range product  Mid-range product   
1                 1   both   both  mid-range product  Mid-range product   
2                 0   both   both  mid-range product  Mid-range product   
3                 0   both   both  mid-range product  Mid-range product   
4                 1   both   both  mid-range product  Mid-range product   
...             ...    ...    ...                ...                ...   
32404854          0   both   both  mid-range product                NaN   
32404855          0   both   both  mid-range product                NaN   
32404856          0   both   both  mid-range product                NaN   
32404857          1   both   both  mid-range product                NaN   
32404858          0   both   both  mid-range product                NaN   

             busiest day  
0            regular day  
1            regular day  
2            busiest day  
3            regular day  
4         least busy day  
...                  ...  
32404854     busiest day  
32404855  least busy day  
32404856     regular day  
32404857     regular day  
32404858     regular day  

[32404859 rows x 18 columns]

# Beginning of Task 4.7

In [108]:
# Identifying days with the most orders
df_ord_prod_merge['order_day_of_week'].value_counts(dropna = False)

order_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [120]:
# amending busiest and least busiest day definitions
result = []

for value in df_ord_prod_merge['order_day_of_week']:
    if value == 0 or value == 1:
        result.append('busiest days')
    elif value == 4 or value == 3:
        result.append('least busy days')
    else:
        result.append('regular day')

In [121]:
df_ord_prod_merge['busiest days'] = result

In [130]:
# New value counts for new labels
df_ord_prod_merge['busiest days'].value_counts(dropna = False)

busiest days
regular day        12916111
busiest days       11864412
least busy days     7624336
Name: count, dtype: int64

For comparison, below are the previous value counts before adding Sunday and Tuesday to the busiest and slowest days, respectively.

busiest day
regular day       22416875
busiest day        6204182
least busy day     3783802
Name: count, dtype: int64

Redefining the days with the most and least orders changed those labels pretty significantly.

In [132]:
df_ord_prod_merge = df_ord_prod_merge.drop('busiest day', axis=1)

In [134]:
df_ord_prod_merge

product_id                product_name  aisle_id  department_id  \
0                  1  Chocolate Sandwich Cookies        61             19   
1                  1  Chocolate Sandwich Cookies        61             19   
2                  1  Chocolate Sandwich Cookies        61             19   
3                  1  Chocolate Sandwich Cookies        61             19   
4                  1  Chocolate Sandwich Cookies        61             19   
...              ...                         ...       ...            ...   
32404854       49688      Fresh Foaming Cleanser        73             11   
32404855       49688      Fresh Foaming Cleanser        73             11   
32404856       49688      Fresh Foaming Cleanser        73             11   
32404857       49688      Fresh Foaming Cleanser        73             11   
32404858       49688      Fresh Foaming Cleanser        73             11   

          prices  order_id  user_id  order_number  order_day_of_week  \
0            5.8   3139998      138            28                  6   
1            5.8   1977647      138            30                  6   
2            5.8    389851      709             2                  0   
3            5.8    652770      764             1                  3   
4            5.8   1813452      764             3                  4   
...          ...       ...      ...           ...                ...   
32404854    13.5   1788356   200215             2                  0   
32404855    13.5   3401313   200377             1                  4   
32404856    13.5    809510   200873             5                  3   
32404857    13.5   2359893   200873             9                  3   
32404858    13.5   2385091   205926            11                  1   

          order_hour_of_day  days_since_last_order  add_to_cart_order  \
0                        11                    3.0                  5   
1                        17                   20.0                  1   
2                        21                    6.0                 20   
3                        13                    NaN                 10   
4                        17                    9.0                 11   
...                     ...                    ...                ...   
32404854                  9                    5.0                 27   
32404855                 11                    NaN                  5   
32404856                  8                   15.0                 12   
32404857                 15                    5.0                 11   
32404858                 15                    6.0                 35   

          reordered _merge source        price_range    price_range_loc  \
0                 0   both   both  mid-range product  Mid-range product   
1                 1   both   both  mid-range product  Mid-range product   
2                 0   both   both  mid-range product  Mid-range product   
3                 0   both   both  mid-range product  Mid-range product   
4                 1   both   both  mid-range product  Mid-range product   
...             ...    ...    ...                ...                ...   
32404854          0   both   both  mid-range product                NaN   
32404855          0   both   both  mid-range product                NaN   
32404856          0   both   both  mid-range product                NaN   
32404857          1   both   both  mid-range product                NaN   
32404858          0   both   both  mid-range product                NaN   

             busiest days  
0             regular day  
1             regular day  
2            busiest days  
3         least busy days  
4         least busy days  
...                   ...  
32404854     busiest days  
32404855  least busy days  
32404856  least busy days  
32404857  least busy days  
32404858     busiest days  

[32404859 rows x 18 columns]

I dropped the "busiest day" column so there won't be any confusion.

In [139]:
# Identifying busiest times of the day
df_ord_prod_merge['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

In [163]:
# Assigning labels for each hour of the day based on order volume
busyhour = []

for value in df_ord_prod_merge['order_hour_of_day']:
    if value in [10, 11, 14, 15, 13, 12, 16, 9]:
        busyhour.append('Most Orders')
    elif value in [17, 8, 18, 19, 20, 7, 21, 22]:
        busyhour.append('Average Orders')
    elif value in [23, 6, 0, 1, 2, 3, 4, 5]:
        busyhour.append('Fewest Orders')

In [171]:
# adding the new column to the dataframe
df_ord_prod_merge['busiest_period_of_day'] = busyhour

In [173]:
# Checking value counts for new column
df_ord_prod_merge['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Most Orders       21118071
Average Orders     9997651
Fewest Orders      1289137
Name: count, dtype: int64

In [169]:
df_ord_prod_merge

product_id                product_name  aisle_id  department_id  \
0                  1  Chocolate Sandwich Cookies        61             19   
1                  1  Chocolate Sandwich Cookies        61             19   
2                  1  Chocolate Sandwich Cookies        61             19   
3                  1  Chocolate Sandwich Cookies        61             19   
4                  1  Chocolate Sandwich Cookies        61             19   
...              ...                         ...       ...            ...   
32404854       49688      Fresh Foaming Cleanser        73             11   
32404855       49688      Fresh Foaming Cleanser        73             11   
32404856       49688      Fresh Foaming Cleanser        73             11   
32404857       49688      Fresh Foaming Cleanser        73             11   
32404858       49688      Fresh Foaming Cleanser        73             11   

          prices  order_id  user_id  order_number  order_day_of_week  \
0            5.8   3139998      138            28                  6   
1            5.8   1977647      138            30                  6   
2            5.8    389851      709             2                  0   
3            5.8    652770      764             1                  3   
4            5.8   1813452      764             3                  4   
...          ...       ...      ...           ...                ...   
32404854    13.5   1788356   200215             2                  0   
32404855    13.5   3401313   200377             1                  4   
32404856    13.5    809510   200873             5                  3   
32404857    13.5   2359893   200873             9                  3   
32404858    13.5   2385091   205926            11                  1   

          order_hour_of_day  days_since_last_order  add_to_cart_order  \
0                        11                    3.0                  5   
1                        17                   20.0                  1   
2                        21                    6.0                 20   
3                        13                    NaN                 10   
4                        17                    9.0                 11   
...                     ...                    ...                ...   
32404854                  9                    5.0                 27   
32404855                 11                    NaN                  5   
32404856                  8                   15.0                 12   
32404857                 15                    5.0                 11   
32404858                 15                    6.0                 35   

          reordered _merge source        price_range    price_range_loc  \
0                 0   both   both  mid-range product  Mid-range product   
1                 1   both   both  mid-range product  Mid-range product   
2                 0   both   both  mid-range product  Mid-range product   
3                 0   both   both  mid-range product  Mid-range product   
4                 1   both   both  mid-range product  Mid-range product   
...             ...    ...    ...                ...                ...   
32404854          0   both   both  mid-range product                NaN   
32404855          0   both   both  mid-range product                NaN   
32404856          0   both   both  mid-range product                NaN   
32404857          1   both   both  mid-range product                NaN   
32404858          0   both   both  mid-range product                NaN   

             busiest days     busiest day busiest_period_of_day  
0             regular day     regular day           Most Orders  
1             regular day     regular day        Average Orders  
2            busiest days     busiest day        Average Orders  
3         least busy days     regular day           Most Orders  
4         least busy days  least busy day        Average Orders  
...                   ...             ...     

In [179]:
df_ord_prod_merge.to_pickle('/Users/samabrams/Data Analysis Projects/Instacart Basket Analysis/02 Data/Prepared Data/orders_products_merged_w_labels.pkl')